### ISSUE TITLE
Adding Z-Loss

### ISSUE THREAD

**Issue by francesco-bertolotti** [pytorch/torchtitan#2523]
The **z-loss** is a commonly used technique to prevent logit explosion, and it is currently used in both OLMo3 and Trinity. Would you be interested in adding support for it in `torchtitan`, perhaps in [loss.py](https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/loss.py), with a configurable weight?

It could also be useful to log the z-loss alongside the NTP loss.

To support this and similar extensions, we could modify the loss interface so that it returns either a dictionary or a custom loss object that tracks both the main loss and any auxiliary losses.

---

If this sounds useful to you as well, I’d be happy to start working on a PR implementing this feature.

**Comment by tianyu-l** [comment]
Thanks. Is it fair to say that z-loss has not become the de facto standard in llm training?

If so, I think we can start with adding z-loss to logging (in no-grad mode), but not using it as a default regularizer.

Whether we can make it configurable to turn on depends on how the implementation looks, which we can discuss after you have one. WDYT?

cc @felipemello1 who's looking at improving logging in general.

**Comment by francesco-bertolotti** [comment]
@tianyu-l 

I should be able to submit a PR this week.

**Comment by felipemello1** [comment]
@francesco-bertolotti , regarding the loss, for now you will have to do the manual all_reduce and logging like titan currently does it. After we make some changes to make titan for dtensor native, you should be able to do something like: `record_tensor_metric(sum=loss.sum(), count=(mask==1).sum())`, and we will handle the all reduce and logging for you.

That said, I am about to push a PR changing MetricsProcessor greatly. So perhaps just add the z-loss and print or log it naively, but dont make changes to the MetricsProcessor. You will have many conflicts. Then you can rebase on my PR.
